src/lib/io/sol-fs-storage.c

```diff
@@ -41,35 +41,95 @@
 
 #include "sol-buffer.h"
 #include "sol-log.h"
+#include "sol-mainloop.h"
 #include "sol-util.h"
 #include "sol-util-file.h"
 
+struct cb_data {
+    char *name;
+    struct sol_blob *blob;
+    void (*cb)(void *data, const char *name, struct sol_blob *blob, int status);
+    const void *data;
+    int status;
+};
+
+static bool
+write_cb(void *data)
+{
+    struct cb_data *cb_data = data;
+
+    if (cb_data->cb)
+        cb_data->cb((void *)cb_data->data, cb_data->name, cb_data->blob,
+            cb_data->status);
+    sol_blob_unref(cb_data->blob);
+
+    free(cb_data->name);
+    free(cb_data);
+
+    return false;
+}
+
 SOL_API int
-sol_fs_write_raw(const char *name, const struct sol_buffer *buffer)
+sol_fs_write_raw(const char *name, struct sol_blob *blob,
+    void (*cb)(void *data, const char *name, struct sol_blob *blob, int status),
+    const void *data)
 {
-    FILE *file;
-    int ret = 0;
+    FILE *file = NULL;
+    struct cb_data *cb_data;
+    int ret = 0, status = 0;
 
     SOL_NULL_CHECK(name, -EINVAL);
-    SOL_NULL_CHECK(buffer, -EINVAL);
+    SOL_NULL_CHECK(blob, -EINVAL);
+
+    if (cb) {
```

Comment:
I don't like this approach of different execution flows if a callback is given or not.

For these async callbacks I'd force user to provide a callback, otherwise he will be missing error handling. If he wants to ignore error handling, he will have to provide a dummy callback -- likely he will take the time and handle the error code :-)

in every way, please execute the write from the timer. It will also make it easier in future versions to move this to a worker thread... which reminds me: you should be keeping an array of pending writes and cancel the old if a new one comes. While keeping the timer, it will not be much of a problem, but if you add threads one could step on each other foot!